// @japa/core/build/src/suite/runner.js
"use strict";
/*
 * @japa/core
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuiteRunner = void 0;
const debug_1 = __importDefault(require("../debug"));
/**
 * Run all groups or tests inside the suite stack
 */
class SuiteRunner {
    constructor(suite, hooks, emitter) {
        this.suite = suite;
        this.hooks = hooks;
        this.emitter = emitter;
        /**
         * Reference to the startup runner
         */
        this.setupRunner = this.hooks.runner('setup');
        /**
         * Reference to the cleanup runner
         */
        this.teardownRunner = this.hooks.runner('teardown');
        /**
         * Test errors
         */
        this.errors = [];
        /**
         * Track if test has any errors
         */
        this.hasError = false;
    }
    /**
     * Notify the reporter about the suite start
     */
    notifyStart() {
        const startOptions = { name: this.suite.name };
        this.emitter.emit('suite:start', startOptions);
    }
    /**
     * Notify the reporter about the suite end
     */
    notifyEnd() {
        const endOptions = {
            name: this.suite.name,
            hasError: this.hasError,
            errors: this.errors,
        };
        this.emitter.emit('suite:end', endOptions);
    }
    /**
     * Running setup hooks
     */
    async runSetupHooks() {
        (0, debug_1.default)('running "%s" suite setup hooks', this.suite.name);
        try {
            await this.setupRunner.run(this.suite);
        }
        catch (error) {
            (0, debug_1.default)('suite setup hooks failed, suite: %s, error: %O', this.suite.name, error);
            this.hasError = true;
            this.errors.push({ phase: 'setup', error });
        }
    }
    /**
     * Running teardown hooks
     */
    async runTeardownHooks() {
        (0, debug_1.default)('running "%s" suite teardown hooks', this.suite.name);
        try {
            await this.teardownRunner.run(this.suite);
        }
        catch (error) {
            (0, debug_1.default)('suite teardown hooks failed, suite: %s, error: %O', this.suite.name, error);
            this.hasError = true;
            this.errors.push({ phase: 'teardown', error });
        }
    }
    /**
     * Running setup cleanup functions
     */
    async runSetupCleanupFunctions() {
        (0, debug_1.default)('running "%s" suite setup cleanup functions', this.suite.name);
        try {
            await this.setupRunner.cleanup(this.hasError, this.suite);
        }
        catch (error) {
            (0, debug_1.default)('suite setup cleanup functions failed, suite: %s, error: %O', this.suite.name, error);
            this.hasError = true;
            this.errors.push({ phase: 'setup:cleanup', error });
        }
    }
    /**
     * Running teardown cleanup functions
     */
    async runTeardownCleanupFunctions() {
        (0, debug_1.default)('running "%s" suite teardown cleanup functions', this.suite.name);
        try {
            await this.teardownRunner.cleanup(this.hasError, this.suite);
        }
        catch (error) {
            (0, debug_1.default)('suite teardown cleanup functions failed, suite: %s, error: %O', this.suite.name, error);
            this.hasError = true;
            this.errors.push({ phase: 'teardown:cleanup', error });
        }
    }
    /**
     * Run the test
     */
    async run() {
        (0, debug_1.default)('starting to run "%s" suite', this.suite.name);
        this.notifyStart();
        /**
         * Run setup hooks and exit early when one of the hooks
         * fails
         */
        await this.runSetupHooks();
        if (this.hasError) {
            await this.runSetupCleanupFunctions();
            this.notifyEnd();
            return;
        }
        /**
         * Run the test executor
         */
        for (let groupOrTest of this.suite.stack) {
            await groupOrTest.exec();
        }
        /**
         * Cleanup setup hooks
         */
        await this.runSetupCleanupFunctions();
        /**
         * Run + cleanup teardown hooks
         */
        await this.runTeardownHooks();
        await this.runTeardownCleanupFunctions();
        /**
         * Notify test end
         */
        this.notifyEnd();
    }
}
exports.SuiteRunner = SuiteRunner;
