// japa/build/src/Contracts/index.js
"use strict";
/**
 * @module Core
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ITestStatus = exports.IGroupStatus = exports.IEvents = void 0;
var IEvents;
(function (IEvents) {
    IEvents["TESTSTARTED"] = "test:started";
    IEvents["TESTCOMPLETED"] = "test:completed";
    IEvents["GROUPSTARTED"] = "group:started";
    IEvents["GROUPCOMPLETED"] = "group:completed";
    IEvents["STARTED"] = "started";
    IEvents["COMPLETED"] = "completed";
})(IEvents = exports.IEvents || (exports.IEvents = {}));
var IGroupStatus;
(function (IGroupStatus) {
    IGroupStatus["FAILED"] = "failed";
    IGroupStatus["PASSED"] = "passed";
    IGroupStatus["PENDING"] = "pending";
})(IGroupStatus = exports.IGroupStatus || (exports.IGroupStatus = {}));
var ITestStatus;
(function (ITestStatus) {
    ITestStatus["FAILED"] = "failed";
    ITestStatus["PASSED"] = "passed";
    ITestStatus["SKIPPED"] = "skipped";
    ITestStatus["TODO"] = "todo";
    ITestStatus["PENDING"] = "pending";
})(ITestStatus = exports.ITestStatus || (exports.ITestStatus = {}));
